[13.0] Create module stock_location_children

grindtildeath opened the pull request:
This module adds a `children_ids` field on `stock.location` in order to compute
and store all the children for a `stock.location` and not only its first level
children as is the case for `child_ids`

grindtildeath committed:
Create module stock_location_children

simahawk commented on stock_location_children/models/stock_location.py:
```diff
@@ -0,0 +1,27 @@
+
+    @api.depends('child_ids', 'child_ids.children_ids')
+    def _compute_children_ids(self):
+        for loc in self:
```
can't we use `parent_path` to make this more performant?
Can you check its usage in stock module?

grindtildeath committed:
Use SQL to improve performance

grindtildeath committed:
Run pre-commit

simahawk commented on stock_location_children/models/stock_location.py:
```diff
@@ -0,0 +1,44 @@
+        help="All the children (multi-level) stock location of this location",
+    )
+
+    @api.depends("child_ids", "child_ids.children_ids")
```
shouldn't be `child_ids.child_ids`? :thinking:

simahawk replied:
or better `parent_path`

gurneyalex replied:
Test code:

```
>>> Loc = self.env['stock.location']
>>> stock = Loc.browse(8)
>>> t1 = Loc.create({'name': 'test1', 'location_id': stock.id})
>>> t2 = Loc.create({'name': 'test2', 'location_id': t1.id})
>>> t3 = Loc.create({'name': 'test2', 'location_id': t2.id})
>>> self.env.cr.commit()
```
With api.depends('child_ids', 'child_ids.child_ids'), the computation method is called once with self = stock.location(7, 8, 26, 27, 28)

With api.depends('child_ids', 'child_ids.children_ids', the computation method is called 6 times with self containing a single location each time. 

@grindtildeath  My advice: prefer child_ids.child_ids if it does not break any known test case as it seems more efficient. 

@simahawk I'm not sure how to depend on parent_path. Can you suggest something?

simahawk replied:
@gurneyalex thanks for testing, I suspected this behavior.
Regarding `parent_path`: never mind. I did a quick check and I don't think is going to work because it's handled in "magic" methods.

grindtildeath commented on stock_location_children/models/stock_location.py:
```diff
@@ -0,0 +1,44 @@
+        help="All the children (multi-level) stock location of this location",
+    )
+
+    @api.depends("child_ids", "child_ids.children_ids")
```
```suggestion
    @api.depends("child_ids", "child_ids.child_ids")
```

grindtildeath committed:
Update stock_location_children/models/stock_location.py

simahawk committed:
stock_location_children: generate setup folder

grindtildeath commented on stock_location_children/models/stock_location.py:
```diff
@@ -0,0 +1,44 @@
+            AND sub.id IN %s
+            GROUP BY sub.id;
+        """
+        self.env.cr.execute(query, (tuple(self.ids),))
```
```suggestion
        self.flush()
        self.env.cr.execute(query, (tuple(self.ids),))
```

guewen replied:
It should not flush all the fields because it can flush too much (following relations and flushing other models, ...).
`flush("location_id", "child_ids")`?

guewen replied:
Pushed a new commit

Commit:
Ensure children_ids is computed with proper triggers

* Any child_ids changed on a children_ids should recompute locations:
  otherwise only 2 levels are updated.
* Flush any change in the hierarchy before executing the SQL

Commit:
Fix test compatibility with other modules

When other modules add locations in "stock.stock_location_stock", the
assertions miss them. Create a new tree of locations to run the tests.

simahawk commented:
/ocabot merge patch
